File: Fall/DS_6001/inclass2/DataWrangling.R
# Marcus Rosti
# mer3ef
# DS 6001

#*****************************************************************************
#
#					Data Wrangling - Train Accidents
#
#*****************************************************************************

setwd("~/MSDS/Fall/DS_6001/inclass2")

#************************************
# Reading in data
#************************************
l

# Read all the data into one data frame
trim <- function(x)
  gsub("^\\s+|\\s+$", "", x)


# I put all my files into a sub directory so I hd to parse them correctly
data_files <- list.files("data")
datasets = rep("",14)
datasets <- sapply(data_files,function(x)
  paste("data/",x,sep = ""))
all_trains <- sapply(datasets, function(x)
  read.csv(x))
other_trains <- all_trains[-1]


all_sets <-
  Reduce(function(x,y)
    merge(x ,y ,all = TRUE), other_trains, all_trains[[1]])


# To get a summary of all of the variables
summary(all_sets[,c("ACCDMG","TRNSPD","TONS","TOTKLD","TOTINJ")])


# How many accident reports and variables?


##################################################
#
#	Data Cleaning
#
##################################################

#*************************************************
#		Duplicates
#*************************************************


# Are there other duplicates?
sum(duplicated(all_sets$INCDTNO))

# Remove duplicates
all_sets[!duplicated(all_sets$INCDTNO),]

nodup_all_sets <-
  all_sets[!duplicated(all_sets[,c("YEAR","DAY","MONTH","TIMEHR","TIMEMIN")]),]
#*************************************************
#		Exteme Point
#*************************************************



# Look at the most costly accident in the data set

boxplot(nodup_all_sets$ACCDMG)

# Check out the narratives for this extreme accident

nodup_all_sets[which.max(nodup_all_sets$ACCDMG),c("ACCDMG","YEAR","MONTH","DAY","NARR1")]
# it's the 9/11 attack

# what should we do?
no911_all_sets <- nodup_all_sets[-which.max(nodup_all_sets$ACCDMG),]

#********************************************
# Missing data
#********************************************

# Are we missing values for any variables?
# yes!

# How many?


nafind <- function(x) {
  sum(is.na(x))
}

nacount_all_sets <- apply(no911_all_sets,2,"nafind")

sum(apply(no911_all_sets,2,"nafind") > 0)

# Do we need all the variables?

# defintely not

# Remove unnecessary variables, then get a summary
# Keep TYPEQ, we'll use it.

varWna_all_sets <- which(nacount_all_sets > 0)

varWna_all_sets <-
  varWna_all_sets[-which(colnames(no911_all_sets)[varWna_all_sets] == "TYPEQ")]

clean_all_sets <- no911_all_sets[,-varWna_all_sets]

#*************************************************
#
#   Reformating variables
#
#*************************************************


# Type of accident
# put in the values from the variable definitions

summary(clean_all_sets$TYPE)
#clean_2001< factor(clean_2001$TYPE,labels = c("Derailment","HeadsOn","Rearmed","Side","Raking","BrokenTrain","Hwy-Rail",
#                                              "GradeX","Obstruction","Explosive","Fire","Other","SeeNarrative"))
clean_all_sets$TYPE <-
  factor(
    clean_all_sets$TYPE,labels = c(
      "Derailment","HeadsOn","Rearmed","Side","Raking","BrokenTrain","Hwy-Rail",
      "GradeX","Obstruction","Explosive","Fire","Other","SeeNarrative"
    )
  )
summary(clean_all_sets$TYPE)
# Type of Train
# put in the values from the definitions

clean_all_sets$TYPEQ <-
  factor(
    clean_all_sets$TYPEQ,levels = list(
      c("1"),c("2"),c("3"),c("4"),c("5"),c("6"),c("7"),c("8"),c("9"),c("","A","B","C","D","E")
    ),labels = c(
      "Freight","Passenger","Commuter","Work","Single","CutofCars","Yard","Light","Maint","Other"
    )
  )
summary(clean_all_sets$TYPEQ)

# Clean up the values



# New labels



# Don't do it, but think about
# how you can replace the missing values



# Create a new variable called Cause
# that uses the first character in the string labels for cause.

clean_all_sets$Cause <- rep(NA,nrow(clean_all_sets))

clean_all_sets$Cause[which(substr(clean_all_sets$CAUSE,1,1) == "M")] <- "M"
clean_all_sets$Cause[which(substr(clean_all_sets$CAUSE,1,1) == "T")] <- "T"
clean_all_sets$Cause[which(substr(clean_all_sets$CAUSE,1,1) == "S")] <- "S"
clean_all_sets$Cause[which(substr(clean_all_sets$CAUSE,1,1) == "H")] <- "H"
clean_all_sets$Cause[which(substr(clean_all_sets$CAUSE,1,1) == "E")] <- "E"

clean_all_sets$Cause <- factor(clean_all_sets$Cause)

summary(clean_all_sets$Cause)

#*************************************************
#
#   Reformating narratives
#
#*************************************************

# Look at the narratives
# Combine them into one vector
# where each entry is the complete narrative
# for that accident

clean_all_sets$NARR <- rep(NA,nrow(clean_all_sets))

clean_all_sets$NARR <- apply(clean_all_sets[,c(
  "NARR1","NARR2","NARR3","NARR4","NARR5","NARR6","NARR7","NARR8","NARR9","NARR10","NARR11","NARR12","NARR13","NARR14","NARR15"
)],1,paste,collapse = "")

# GG

File: Fall/DS_6001/inclass2/DataWrangling.py
# Data management
import pandas as pd
from pandas import Series, DataFrame
import numpy as np
import os
from collections import Counter

#There's a weird pandas warning that doesn't apply
pd.options.mode.chained_assignment = None

def print_header():
    print("    ___      _          __    __                       _ _             ")
    print("   /   \__ _| |_ __ _  / / /\ \ \_ __ __ _ _ __   __ _| (_)_ __   __ _ ")
    print("  / /\ / _` | __/ _` | \ \/  \/ | '__/ _` | '_ \ / _` | | | '_ \ / _` |")
    print(" / /_/| (_| | || (_| |  \  /\  /| | | (_| | | | | (_| | | | | | | (_| |")
    print("/___,' \__,_|\__\__,_|   \/  \/ |_|  \__,_|_| |_|\__, |_|_|_| |_|\__, |")
    print("                                                 |___/           |___/ ")

# helper function
def print_breaks():
    breaks = ""
    for i in range(81):
        breaks = breaks + "="
    print(breaks)

def read_datafiles(filename):
    return pd.read_csv(filename, encoding="ISO-8859-1",\
    dtype={'user_id':np.float64}, low_memory=False)

def getFiles(directory):
    os.chdir(directory)
    return [f for f in os.listdir('.') if os.path.isfile(f)]

def getDataFrame(files_to_import):
    read_files = map(read_datafiles, files_to_import)
    dataframes = map(DataFrame, read_files)
    return pd.concat(dataframes, ignore_index=False)

def removeDuplicates(df_to_clean):
    return df_to_clean.drop_duplicates\
    (["YEAR", "MONTH", "DAY", "TIMEHR", "TIMEMIN"]).reset_index()

def remove911(df_to_clean):
    (damage, nine_11_index) = max((v, i) for i, v in enumerate(df_to_clean.ACCDMG))
    return df_to_clean.drop([nine_11_index]).reset_index()

def combineNarratives(df_with_sep_narrs):
    numbers = []
    for i in range(1, 16):
        numbers.append("NARR" + str(i))

    ## concat the narratives together functionally
    df_with_combined_narrs = df_with_sep_narrs
    df_with_combined_narrs.NARR = df_with_sep_narrs[numbers].fillna('').sum(1)
    return df_with_combined_narrs

def fixType(df_dirty_type):
    map_type = {1:"Derailment", 2:"HeadsOn", 3:"Rearmed", 4:"Side", 5 :"Raking", \
    6:"BrokenTrain", 7:"Hwy-Rail", 8:"GradeX", 9:"Obstruction", 10:"Explosive", \
    11:"Fire", 12:"Other", 13:"SeeNarrative"}
    df_clean_type = df_dirty_type
    df_clean_type.TYPE = df_dirty_type.TYPE.map(map_type)
    return df_clean_type

def fixTypeQ(df_dirty_levels):
    # new levels
    map_typeq = {1:"Freight", 2:"Passenger", 3:"Commuter", 4:"Work",\
                 5:"Single", 6:"CutofCars", 7:"Yard", 8:"Light", 9:"Maint"}
    df_clean_levels = df_dirty_levels
    df_clean_levels.TYPEQ = df_dirty_levels.TYPEQ.map(map_typeq)
    return df_clean_levels

def dropNAs(df_dirty_data):
    """ Drops NAs - with stipulations

        I thought this required an explanation. It doesn't drop every single na.
        It drops columns with NAs except for TYPE and TYPEQ

        param:
            df_with_all_data - dataframe with columns that contain NAs
        return:
            df_clean_data    - dataframe that has no columns with NAs
    """
    # Save certain columns
    type_temp = df_dirty_data.TYPE
    typeQ_temp = df_dirty_data.TYPEQ

    # Remove anything with an NA
    df_clean_data = df_dirty_data.dropna(axis=1)

    # Add columns back into the df
    df_clean_data.TYPE = type_temp
    df_clean_data.TYPEQ = typeQ_temp

    # Return the clean df
    return df_clean_data

def causeFixerFunction(cause_to_fix):
    """ Functional way to a fix single cause
    """
    if cause_to_fix.startswith("T"):
        return "T"
    if cause_to_fix.startswith("H"):
        return "H"
    if cause_to_fix.startswith("M"):
        return "M"
    if cause_to_fix.startswith("E"):
        return "E"
    else:
        return "S"

def fixCauseInPlace(df_dirty_cause):
    df_clean_cause = df_dirty_cause
    df_clean_cause.CAUSE = df_dirty_cause.CAUSE.map(causeFixerFunction)
    return df_clean_cause

# Entry point
def main():
    print_breaks()
    print_header()
    print_breaks()
    files = []
    init_df = DataFrame()
    print("Reading in data files")
    try: # encapsulate unsafe IO into a try/catch
        # get files to read
        files = getFiles("/Users/RustyRosti/MSDS/Fall/DS_6001/inclass2/data")
        # read files into a dataframe
        init_df = getDataFrame(files)
    except:
        #Error reading in files
        print("There's some IO error")
        return -1

    print(init_df[["ACCDMG", "TRNSPD",\
    "TONS", "CARSDMG", "TOTINJ", "TOTKLD"]].describe())
    print_breaks()

    print("Removing 911")
    print("The other outliers are interesting points")
    no_dups_df = removeDuplicates(init_df)
    no_911_df = remove911(no_dups_df)
    print("The number of rows without dups: ", no_911_df.shape)
    print("The most expensive accident: ", max(no_911_df.ACCDMG))
    print_breaks()

    print("Concatenating Narratives")
    one_narr_df = combineNarratives(no_911_df)
    print(one_narr_df.NARR.describe())
    print_breaks()

    print("Fixing the types")
    fixed_type_df = fixType(one_narr_df)
    print(Counter(fixed_type_df.TYPE))
    print_breaks()

    print("Fixing TypeQ level")
    fixed_levels_df = fixTypeQ(fixed_type_df)
    print(Counter(fixed_levels_df.TYPEQ))
    print_breaks()

    print("Fixing causes")
    fixed_cause_df = fixCauseInPlace(fixed_levels_df)
    print(Counter(fixed_cause_df.CAUSE))
    print_breaks()

    print("Removing NAs")
    no_nas_df = dropNAs(fixed_cause_df)
    print(no_nas_df.describe())
    print_breaks()


    print("Data Wrangled!")
    return 0

if __name__ == "__main__":
    main()
